src/NonfungiblePositionManager.sol

```diff
@@ -0,0 +1,226 @@
+        return abi.encode(delta);
+    }
+
+    function _mint(bytes memory param) internal returns (bytes memory) {
+        (LiquidityRange memory range, uint256 liquidity, address owner, bytes memory hookData) =
+            abi.decode(param, (LiquidityRange, uint256, address, bytes));
+
+        // mint receipt token
+        uint256 tokenId;
+        unchecked {
+            tokenId = nextTokenId++;
+        }
+        _mint(owner, tokenId);
+
+        (BalanceDelta delta,) = _modifyLiquidity(range, liquidity.toInt256(), bytes32(tokenId), hookData);
+
+        tokenRange[tokenId] = range;
+
+        return abi.encode(delta);
+    }
+
+    /// @param params is an encoding of the Currency to close
+    /// @param sender is the msg.sender encoded by the `modifyLiquidities` function before the `unlockCallback`.
+    /// @return int256 the balance of the currency being settled by this call
+    function _close(bytes memory params, address sender) internal returns (bytes memory) {
+        (Currency currency) = abi.decode(params, (Currency));
+        // this address has applied all deltas on behalf of the user/owner
+        // it is safe to close this entire delta because of slippage checks throughout the batched calls.
+        int256 currencyDelta = manager.currencyDelta(address(this), currency);
+
+        // the sender is the payer or receiver
+        if (currencyDelta < 0) {
+            currency.settle(manager, sender, uint256(-int256(currencyDelta)), false);
+        } else {
+            currency.take(manager, sender, uint256(int256(currencyDelta)), false);
+        }
+
+        return abi.encode(currencyDelta);
+    }
+
+    function burn(uint256 tokenId, address sender) internal {
+        _requireApprovedOrOwner(tokenId, sender);
+        // We do not need to enforce the pool manager to be unlocked bc this function is purely clearing storage for the minted tokenId.
+
+        // Checks that the full position's liquidity has been removed and all tokens have been collected from tokensOwed.
+        _validateBurn(tokenId);
+
+        delete tokenRange[tokenId];
+        // Burn the token.
+        _burn(tokenId);
+    }
+
+    function _modifyLiquidity(LiquidityRange memory range, int256 liquidityChange, bytes32 salt, bytes memory hookData)
+        internal
+        returns (BalanceDelta liquidityDelta, BalanceDelta totalFeesAccrued)
+    {
+        (liquidityDelta, totalFeesAccrued) = manager.modifyLiquidity(
+            range.poolKey,
+            IPoolManager.ModifyLiquidityParams({
+                tickLower: range.tickLower,
+                tickUpper: range.tickUpper,
+                liquidityDelta: liquidityChange,
+                salt: salt
+            }),
+            hookData
+        );
+    }
+
+    function _validateBurn(uint256 tokenId) internal {
+        bytes32 positionId = getPositionIdFromTokenId(tokenId);
+        uint128 liquidity = manager.getPositionLiquidity(tokenRange[tokenId].poolKey.toId(), positionId);
```

Comment:
the one thing that irks me right now is how we're reading from tokenRange twice -- once in `getPositionIdFromTokenId` and again in arguments for `getPositionLiquidity`

leaning towards optimizing this later but just wanted to flag

Reply:
good call out, we can definitely optimize. I also would like to move `getPositionIdFromTokenId` into a view only library

Reply:
i have some ideas here - im tracking a library change here with some comments about it. https://github.com/Uniswap/v4-periphery/issues/171